lang/ghc/Portfile

```diff
@@ -29,144 +24,173 @@ long_description    \
+
+    if {[variant_isset "bootstrap"]} {
+        ui_msg "\tBootstrap build/destroot from ${name}-${version}-src${extract.suffix}…"
+        xinstall -m 0755 -d ${workpath}/src
+        system -W ${workpath}/src \
+            "tar xJf ${distpath}/${name}-${version}-src${extract.suffix}"
+        system -W ${workpath}/src/${name}-${version} \
+            "./boot"
+        system -W ${workpath}/src/${name}-${version} \
+            "export PATH=\${PATH}:${destroot}${prefix}/bin ; \
+                ./configure --prefix=${destroot}${prefix}"
+        system -W ${workpath}/src/${name}-${version} \
+            "make -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
+        system -W ${workpath}/src/${name}-${version} \
+            "make install -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
     }
-}
 
-post-activate {
-    set libprefix "${prefix}/lib/${name}-${version}"
-
-    # delete old ${prefix}/lib/ghc-${version} directories
-    foreach fullpath [glob -nocomplain -directory ${prefix}/lib ghc-*] {
-        if {$fullpath ne $libprefix} {
-            ui_info "Cleaning up remnant GHC library directory ${fullpath}"
-            catch {delete $fullpath}
+    # recursively remove any destroot path in all destrooted files
+    ui_msg "\tRepairing \${prefix} paths…"
+    proc ::findFiles { baseDir pattern } {
+        set files {}
+        lappend files {*}[ glob -nocomplain -type f [ file join ${baseDir} ${pattern} ] ]
+        foreach dir [ glob -nocomplain -type d [ file join ${baseDir} * ] ] {
+            lappend files {*}[ findFiles ${dir} ${pattern} ]
         }
+        return ${files}
     }
-
-    # Walk the current package.conf.d and delete all files not belonging to any
-    # port, since haskell packages now directly drop it in packages.conf.d and
-    # register the file to themselves. The package cache is updated by the
-    # post-activate phase of the packages, which now runs ghc-pkg recache.
-    #
-    # This *will* cause pre-deactivation failures for all ports that still are
-    # correctly installed and would unregister themselves on uninstall.
-    # Unfortunately there is no good way around this now that the contents of
-    # PortGroups are stored together with Portfiles and used for deactivation
-    # (otherwise I could just edit the PortGroup and make unregistration
-    # failures non-fatal)
-    foreach configfile [glob -nocomplain -directory $libprefix/package.conf.d *] {
-        if {[registry_file_registered $configfile] == 0} {
-            # file is not registered to any port, remove it
-            catch {delete $configfile}
-        }
+    foreach destroot_file [ findFiles ${destroot}${prefix} "*" ] {
+        # delete any destroot path appearing in text files
+        system "\[\[ `/usr/bin/file --mime ${destroot_file} | /usr/bin/egrep -c -e ':\[\[:space:\]\]+text/'` != \"0\" \]\] \
+            && /usr/bin/sed -i '' 's|${destroot}||g' ${destroot_file} \
+            || /usr/bin/true"
```

Comment:
This code looks just about horrible. Is there really no way to achieve the same in some more elegant way? If upstream doesn't support DESTDIR installation, I would file a bug. If this hackery is needed just to avoid a deficiency in the way we build the software, maybe there's a way we could build it in a different way?

Reply:
Not horrible, practical.

The ghc build code allows you to specify an install destination, but then hard codes the install path in all its shell scripts.

I can’t tell ghc to install to ${prefix} because then the install won’t be destrootred.

So I install to ${destroot}${prefix} and fix paths after destrooting.

I can complain about this upstream, but the ghc folks aren’t guaranteed to agree that this is a bug, and who knows if and when it would be addressed.

If there’s a better tcl/Portfile approach than the one proposed here, I’m open to it, but am not aware of one.

Reply:
In the meantime I’ll open an issue upstream and ask if there’s a way to avoid this hack.

FWIW, I did my best to avoid this by editing the configured Makefile options in ghc’s `mk` directory, but ran into the same issue so I took the practical road.

Reply:
There's no issues handler at https://github.com/ghc/ghc. How ghc handles destroot or not is beyond scope.